### ISSUE TITLE
Loss Mask to Prevent Weight Updates for Specific Generated Tokens.

### ISSUE THREAD

**Issue by naphatkps** [huggingface/transformers#33369]
### Feature request

We are giving it a text during fine-tuning, but it includes context and a user question with an expected answer. After the model then predicts, during the calculation of loss, it calculates the loss on the whole given input-context, question, and answer. And here I want to apply a loss mask so that it only backpropagates and updates the weights for the loss from the generated answer and not for context and question.

### Motivation

This loss mask feature would ensure that only the loss from the generated answer is backpropagated, excluding the context and question for efficiency in training. It would make the model more focused on generating answers, hence generalizing and avoiding overfitting to the irrelevant parts of the input sequence.

### Your contribution

I am willing to test this feature and report any bugs I encounter during the process.

**Comment by zucchini-nlp** [comment]
Hey! You can also prepare labels yourself with custom data collators, and set to `-100` the masked part and pad ids. CE loss will ignore the `-100` then

**Comment by NielsRogge** [comment]
Indeed, the TRL library for instance has the `DataCollatorForCompletionOnlyLM` class for that purpose: https://huggingface.co/docs/trl/en/sft_trainer#train-on-completions-only.

**Comment by naphatkps** [comment]
Thank you so much for your help! I will go through the details of the `DataCollatorForCompletionOnlyLM` and try to apply it into my project.
